### DISCUSSION TITLE
About the Slayer CNN definition.

### DISCUSSION THREAD

**Discussion by ahmetakman** [lava-nc/lava-dl#60] [Q&A]
Hello,
I am curious about how exactly slayer CNNs work. As far as I understood, it defines spiking neuron-specific parameters then passes the data to Conv3d of PyTorch. So, in that sense, the data should be like a simple frame, right? Asynchronous event streams are not considered in this fashion if I did not misunderstand. Can someone give a more explicit explanation of Convolutional feature learning in SLAYER?

**Comment by alexggener** [comment]
Hi, I'm interested as well on this topic. 

I'm trying to replicate the SNN network SpikeMS (https://github.com/prgumd/SpikeMS) developed in SLAYERpytorch with the lava-dl. The SpikeMS implementation contains 6 Convolutional layers that process spikes in the following Tensor format
``Tensor=[n_channels, height, width, num_time_bins ]`` as in NMNIST dataset lava-dl provides.

The current SNN implementation I have for the moment is the following:
````
self.blocks = torch.nn.ModuleList([

        slayer.block.cuba.Conv(neuron_params_drop_conv1,  2, 16, 3, 2, 0, dilation=1, groups=1, weight_scale=1),
        slayer.block.cuba.Conv(neuron_params_drop_conv1, 16, 32, 3, 2, 0, dilation=1, groups=1, weight_scale=1),
        slayer.block.cuba.Conv(neuron_params_drop_conv1, 32, 64, 3, 2, 0, dilation=1, groups=1, weight_scale=1),
        slayer.block.cuba.Conv(neuron_params_drop_conv1, 64, 32, 3, 2, 0, dilation=1, groups=1, weight_scale=1),
        slayer.block.cuba.Conv(neuron_params_drop_conv1, 32, 16, 3, 2, 0, dilation=1, groups=1, weight_scale=1),
        slayer.block.cuba.Conv(neuron_params_drop_conv1, 16,  2, 3, 2, 0, dilation=1, groups=1, weight_scale=1),
````
And my input spike tensor is ``input_spikes = Tensor[bach, n_channels*height*width, num_time_bins]``. Of course this does not work since dimensions of the ``input_spikes`` and the first layer are not the same, so I've tried to add a dummy Dense Layer as follows
````
self.blocks = torch.nn.ModuleList([
        # Input dummy layer
        slayer.block.cuba.Dense(neuron_input_params_drop, 2*144*256, 16*2*3*3), # channels=2, height=144, width=256
        # Autoencoder layers
        slayer.block.cuba.Conv(neuron_params_drop_conv1,  2, 16, 3, 2, 0, dilation=1, groups=1, weight_scale=1),
        slayer.block.cuba.Conv(neuron_params_drop_conv2, 16, 32, 3, 2, 0, dilation=1, groups=1, weight_scale=1),
        slayer.block.cuba.Conv(neuron_params_drop_conv3, 32, 64, 3, 2, 0, dilation=1, groups=1, weight_scale=1),
        slayer.block.cuba.Conv(neuron_params_drop_conv4, 64, 32, 3, 2, 0, dilation=1, groups=1, weight_scale=1),
        slayer.block.cuba.Conv(neuron_params_drop_conv5, 32, 16, 3, 2, 0, dilation=1, groups=1, weight_scale=1),
        slayer.block.cuba.Conv(neuron_params_drop_conv6, 16,  2, 3, 2, 0, dilation=1, groups=1, weight_scale=1),
````
However, dimensions do not fit between Dense layer and Conv layer.

So any extra help with how CuBa conv blocks work and what is the proper way of using them would be really helpful,

Thanks,
Alex.

**Comment by bamsumit** [comment]
Repeating the [fundamental practice](https://github.com/lava-nc/lava-dl/blob/main/src/lava/lib/dl/slayer/README.md#fundamental-practices) in lava-dl SLAYER: 
> Tensors are always assumed to be in the order `NCHWT` or `NCT` where `N`:Batch, `C`:Channel, `H`: Height(y), `W`: Width(x) and `T`: Time.
`NCHW` is the default PyTorch ordering.

A conv (conv2d) always requires Channel, Height and Width dimension. Pointing you to [conv synapse documentation](https://lava-nc.org/lava-lib-dl/slayer/synapse/synapse.html#lava.lib.dl.slayer.synapse.layer.Conv.forward) (Perhaps this should also be explained at the block documentation), slayer conv expects input in `NCHWT` order.

So @alexggener, you should reshape (or unflatten assuming you have flattened your input) your input tensor to `(batch, 2, 144, 256, num_time_bins)` before you feed it to convolution.

**Reply by alexggener** [reply]
Thanks for your answer. I agree with you that it could be added at the block documentation since they are the high level functions novel users see when they start working with lava-dl. 

Again, thanks for your fast response :) 

Alex.

**Reply by bamsumit** [reply]
Thanks for your feedback. I will create a PR for it.
